File: mini-variational-principle/src/field_theory.c
/**
 * field_theory.c -- Field Theory Variational Principles (L6-L8)
 * Ginzburg-Landau, phase field (Allen-Cahn, Cahn-Hilliard),
 * elastic energy, electromagnetic action.
 */
#include "field_theory.h"
#include <math.h>
#include <stdlib.h>
#include <string.h>
#include <float.h>
#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

double ginzburg_landau_free_energy(const double *psi, const double *dp,
    double a, double b, double c, double x0, double x1, int n)
{
    if (!psi||!dp||n<2) return 0.0;
    double F=0.0,dx=(x1-x0)/(double)(n-1);
    for (int i=0;i<n;i++){
        double p=psi[i],d=dp[i],w=(i==0||i==n-1)?0.5:1.0;
        F+=w*(a*p*p+0.5*b*p*p*p*p+c*d*d)*dx;
    }
    return F;
}

void solve_ginzburg_landau(double a, double b, double c, double L,
    double bcL, double bcR, int nx, int ni, double omega,
    double *x, double *psi)
{
    if (!x||!psi||nx<3) return;
    double dx=L/(double)(nx-1);
    for (int i=0;i<nx;i++){ x[i]=(double)i*dx; psi[i]=0.5*(bcL+bcR); }
    psi[0]=bcL; psi[nx-1]=bcR;
    for (int it=0;it<ni;it++){
        double *po=malloc((size_t)nx*sizeof(double));
        if (!po) return;
        memcpy(po,psi,(size_t)nx*sizeof(double));
        double md=0.0;
        for (int i=1;i<nx-1;i++){
            double lap=(psi[i+1]-2.0*psi[i]+psi[i-1])/(dx*dx);
            double rhs=(2.0*psi[i]+a*psi[i]+b*psi[i]*psi[i]*psi[i])/c;
            double denom=2.0/(dx*dx)+a/c; if(denom<1e-10) denom=1e-10;
            double np=(1.0-omega)*psi[i]+omega*(lap+rhs)/denom;
            double d=fabs(np-psi[i]); if (d>md) md=d; psi[i]=np;
        }
        free(po);
        if (md<1e-10) break;
    }
}

double gl_coherence_length(double a, double c)
{ return (fabs(a)<1e-15)?INFINITY:sqrt(c/fabs(a)); }

double gl_critical_field(double a, double b)
{ return (b<1e-15)?INFINITY:sqrt(4.0*M_PI*a*a/b); }

void allen_cahn_rhs(const double *phi, double eps, double M,
    double dx, int n, double *dp)
{
    if (!phi||!dp||n<3) return;
    for (int i=1;i<n-1;i++){
        double lap=(phi[i+1]-2.0*phi[i]+phi[i-1])/(dx*dx);
        double fp=phi[i]*(phi[i]*phi[i]-1.0);
        dp[i]=M*(eps*eps*lap-fp);
    }
    dp[0]=0.0; dp[n-1]=0.0;
}

void cahn_hilliard_rhs(const double *phi, double eps, double M,
    double dx, int n, double *dp)
{
    if (!phi||!dp||n<5) return;
    double *mu=malloc((size_t)n*sizeof(double));
    if (!mu) return;
    for (int i=1;i<n-1;i++){
        double lap=-(phi[i+1]-2.0*phi[i]+phi[i-1])/(dx*dx);
        double fp=phi[i]*(phi[i]*phi[i]-1.0);
        mu[i]=eps*eps*lap+fp;
    }
    mu[0]=mu[1]; mu[n-1]=mu[n-2];
    for (int i=2;i<n-2;i++)
        dp[i]=M*(mu[i+1]-2.0*mu[i]+mu[i-1])/(dx*dx);
    dp[0]=dp[1]=dp[2]; dp[n-1]=dp[n-2]=dp[n-3];
    free(mu);
}

double interface_energy(const double *phi, double eps, double dx, int n)
{
    if (!phi||n<2) return 0.0;
    double E=0.0;
    for (int i=1;i<n;i++){
        double g=(phi[i]-phi[i-1])/dx, pm=0.5*(phi[i]+phi[i-1]);
        double f=(pm*pm-1.0)*(pm*pm-1.0)*0.25;
        E+=(0.5*eps*eps*g*g+f)*dx;
    }
    return E;
}

double helmholtz_free_energy(const double *strain, const double *C,
    double alpha, double dT, double lam, double mu)
{
    if (!strain||!C) return 0.0;
    double W=0.0;
    for (int i=0;i<6;i++) for (int j=0;j<6;j++) W+=0.5*strain[i]*C[i*6+j]*strain[j];
    double trE=strain[0]+strain[1]+strain[2];
    return W-(3.0*lam+2.0*mu)*alpha*dT*trE;
}

double gibbs_free_energy(const double *stress, const double *S)
{
    if (!stress||!S) return 0.0;
    double G=0.0;
    for (int i=0;i<6;i++) for (int j=0;j<6;j++) G-=0.5*stress[i]*S[i*6+j]*stress[j];
    return G;
}

double em_action_density(const double *E, const double *B,
    double rho, const double *J, const double *A, double phi)
{
    double F2=0.0;
    for (int i=0;i<3;i++){ F2+=E[i]*E[i]-B[i]*B[i]; }
    double JdotA=0.0;
    if (J&&A) for (int i=0;i<3;i++) JdotA+=J[i]*A[i];
    return -0.25*F2-rho*phi+JdotA;
}

void stress_energy_tensor(
    double (*Ld)(const double *f, const double *g, int n),
    const double *f, const double *g, int n, double *T)
{
    if (!Ld||!f||!g||!T||n<1) return;
    double L0=Ld(f,g,n);
    for (int mu=0;mu<4;mu++){
        for (int nu=0;nu<4;nu++){
            double sum=0.0;
            for (int i=0;i<n;i++)
                sum+=g[mu*n+i]*g[nu*n+i]; /* simplified */
            T[mu*4+nu]=sum-(mu==nu?L0:0.0);
        }
    }
}

double topological_charge_density(const double *phi, double dx, int n)
{
    if (!phi||n<3) return 0.0;
    double Q=0.0;
    for (int i=1;i<n;i++){
        double dphi=(phi[i]-phi[i-1])/dx;
        Q+=dphi*dx;
    }
    return Q/(2.0*M_PI);
}

File: mini-variational-principle/src/field_theory.jl
# field_theory.jl — 场论变分原理: Ginzburg-Landau, 相场, 弹性场
# 参考: Landau Vol.5, Ginzburg & Landau 1950

using LinearAlgebra

# ============================================================
# Ginzburg-Landau 自由能
# ============================================================
"""
Ginzburg-Landau 自由能泛函 (1D):
F[ψ] = ∫ [a|ψ|² + b|ψ|⁴/2 + c|∂ψ/∂x|²] dx

a = α(T-Tc), b,c > 0

Euler-Lagrange: c ∂²ψ/∂x² - aψ - b|ψ|²ψ = 0
"""
function ginzburg_landau_free_energy(psi::Function, dpsidx::Function, a::Float64, b::Float64, c::Float64, x_range::Tuple{Float64,Float64}; n=200)
    x0,x1=x_range; dx=(x1-x0)/n; F=0.0
    for i in 1:n
        x=x0+(i-0.5)*dx; p=psi(x); dp=dpsidx(x)
        F += (a*p^2 + 0.5*b*p^4 + c*dp^2)*dx
    end; F
end

"""
Ginzburg-Landau 方程的数值解 (松弛法)
"""
function solve_ginzburg_landau(a::Float64, b::Float64, c::Float64, L::Float64, bc_left::Float64, bc_right::Float64; nx=500, n_iter=5000, omega=1.5)
    dx=L/(nx-1); psi=zeros(nx)
    for i in 1:nx; psi[i]=0.5*(bc_left+bc_right); end
    psi[1]=bc_left; psi[nx]=bc_right
    for _ in 1:n_iter
        psi_old=copy(psi)
        for i in 2:nx-1
            psi[i]=(1-omega)*psi[i]+omega*( (psi[i+1]+psi[i-1])/dx^2 )/( 2/dx^2 + a/c + b*psi[i]^2/c )
        end
        norm(psi-psi_old)<1e-10 && break
    end; range(0,L,length=nx), psi
end

"""
相干长度: ξ = √(c/|a|). 描述序参量的空间变化尺度
"""
gl_coherence_length(a::Float64, c::Float64) = sqrt(c/abs(a))

"""
临界磁场: Hc² = 4π a²/b (type-I超导体热力学临界场)
"""
gl_critical_field(a::Float64, b::Float64) = sqrt(4*pi*a^2/b)

# ============================================================
# 相场模型 (Phase Field)
# ============================================================
"""
Allen-Cahn 方程: ∂φ/∂t = -M δF/δφ = M(ε²∇²φ - f'(φ))

F[φ] = ∫ [ε²/2|∇φ|² + f(φ)] dx
f(φ) = (φ²-1)²/4 (双井势)
"""
function allen_cahn_rhs(phi::Vector{Float64}, epsilon::Float64, M::Float64, dx::Float64)
    n=length(phi); dphi=zeros(n)
    for i in 2:n-1
        laplacian = (phi[i+1]-2*phi[i]+phi[i-1])/dx^2
        f_prime = phi[i]*(phi[i]^2 - 1)  # ∂/∂φ[(φ²-1)²/4] = φ(φ²-1)
        dphi[i] = M*(epsilon^2*laplacian - f_prime)
    end; dphi
end

"""
Cahn-Hilliard 方程 (保守序参量): ∂φ/∂t = ∇·[M∇(δF/δφ)]
"""
function cahn_hilliard_rhs(phi::Vector{Float64}, epsilon::Float64, M::Float64, dx::Float64)
    n=length(phi); mu=zeros(n); dphi=zeros(n)
    for i in 2:n-1
        f_prime = phi[i]*(phi[i]^2 - 1)
        mu[i] = -epsilon^2*(phi[i+1]-2*phi[i]+phi[i-1])/dx^2 + f_prime
    end
    for i in 3:n-2
        dphi[i] = M*(mu[i+1]-2*mu[i]+mu[i-1])/dx^2
    end; dphi
end

"界面能: γ = ∫ ε²/2|∇φ|² + f(φ) dx"
function interface_energy(phi::Vector{Float64}, epsilon::Float64, dx::Float64)
    n=length(phi); E=0.0
    for i in 2:n
        grad = (phi[i]-phi[i-1])/dx; phi_mid = 0.5*(phi[i]+phi[i-1])
        f_val = (phi_mid^2 - 1)^2/4
        E += (0.5*epsilon^2*grad^2 + f_val)*dx
    end; E
end

# ============================================================
# 弹性场变分原理
# ============================================================
"Helmholtz 自由能: Ψ(ε,T) = ½ε:C:ε - (3λ+2μ)α(T-T₀)trε"
function helmholtz_free_energy(strain::Vector{Float64}, C::Matrix{Float64}, alpha::Float64, delta_T::Float64, lam::Float64, mu::Float64)
    0.5*dot(strain, C*strain) - (3*lam+2*mu)*alpha*delta_T*sum(strain[1:3])
end

"Gibbs 自由能: G(σ,T) = -½σ:S:σ"
function gibbs_free_energy(stress::Vector{Float64}, S::Matrix{Float64})
    -0.5*dot(stress, S*stress)
end

# ============================================================
# 电磁场变分原理
# ============================================================
"""
电磁场作用量 (Lorentz 不变):
S[A_μ] = -¼∫ F_{μν}F^{μν} d⁴x + ∫ j^μ A_μ d⁴x

Euler-Lagrange → Maxwell 方程: ∂_μ F^{μν} = j^ν
"""
function em_action_density(E::Vector{Float64}, B::Vector{Float64}, rho::Float64, J::Vector{Float64}, A::Vector{Float64}, phi::Float64)
    # -¼ F² + j·A
    -0.25*(dot(E,E) - dot(B,B)) - rho*phi + dot(J, A)
end

export ginzburg_landau_free_energy, solve_ginzburg_landau, gl_coherence_length, gl_critical_field
export allen_cahn_rhs, cahn_hilliard_rhs, interface_energy
export helmholtz_free_energy, gibbs_free_energy
